Added the Mirage Tank auto disguise

Mailaender opened the pull request:
Closes https://github.com/OpenRA/ra2/issues/183

MustaphaTR commented on OpenRA.Mods.RA2/Traits/Mirage.cs:
```diff
@@ -0,0 +1,257 @@
+				if (conditionManager != null && mirageToken == ConditionManager.InvalidConditionToken && !string.IsNullOrEmpty(condition))
+					mirageToken = conditionManager.GrantCondition(self, condition);
+
+				// Sounds shouldn't play if the actor starts cloaked
```
This comment still says cloak.

Mailaender replied:
Removed it all together.

Mailaender commented:
I updated this with a `WithSpriteBody` subclass which indeed simplifies things a lot. Also added a backfall for maps without trees and used the ones defined in `rules.ini`. The distribution of tree types to use a statistical relevant tree is completely overkill.

sorcerer86pt commented:
Tested this merge request. Although the more technical discussion  details still eludes me, the mirage tank disguise works.

abcdefg30 commented:
Can you please squash the commits with the changes you reverted together?

Mailaender commented:
Squashed.

sorcerer86pt commented:
After the squash , mirage tank disguise is still working,

abcdefg30 commented on OpenRA.Mods.RA2/Traits/Mirage.cs:
```diff
@@ -0,0 +1,260 @@
+	}
+
+	[Desc("This actor can appear as a different actor in specific situations." +
+	" The condition is the randomly selected `MirageTarget` actor type.")]
```
This is no longer true.

abcdefg30 commented on OpenRA.Mods.RA2/Traits/Mirage.cs:
```diff
@@ -0,0 +1,260 @@
+			targetTypes = targets.Select(a => a.Actor.Info).ToArray();
+
+			if (!targetTypes.Any() && info.DefaultTargetTypes != null)
+			{
```
Style nit: No braces around a single line statement.

abcdefg30 commented on OpenRA.Mods.RA2/Traits/Mirage.cs:
```diff
@@ -0,0 +1,260 @@
+		protected override void Created(Actor self)
+		{
+			conditionManager = self.TraitOrDefault<ConditionManager>();
+			otherMirages = self.TraitsImplementing<Mirage>()
```
This is unused!

abcdefg30 commented on OpenRA.Mods.RA2/Traits/Render/WithMirageSpriteBody.cs:
```diff
@@ -0,0 +1,70 @@
+
+		void ITick.Tick(Actor self)
+		{
+			if (mirage.ActorType == null)
```
Shouldn't this stop ticking if the trait is disabled instead?

abcdefg30 commented on mods/ra2/sequences/vehicles.yaml:
```diff
@@ -87,6 +87,7 @@ fv:
 mgtk:
 	Inherits@MC: ^MindControllable
 	icon: rtnkicon
+	idle: null
```
What is this for?

Mailaender replied:
This hack is required.

Mailaender replied:
It's a crash I can't get under control:

```
Exception of type `System.InvalidOperationException`: Unit `mgtk` does not have a sequence named `idle`
  at OpenRA.Graphics.SequenceProvider.GetSequence (System.String unitName, System.String sequenceName) [0x0005d] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/Graphics/SequenceProvider.cs:80 
  at OpenRA.Graphics.Animation.GetSequence (System.String sequenceName) [0x00000] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/Graphics/Animation.cs:217 
  at OpenRA.Graphics.Animation.PlaySequence (System.String sequenceName) [0x00000] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/Graphics/Animation.cs:98 
  at OpenRA.Graphics.Animation.PlayRepeating (System.String sequenceName) [0x0000e] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/Graphics/Animation.cs:106 
  at OpenRA.Mods.Common.Traits.Render.WithSpriteBody..ctor (OpenRA.ActorInitializer init, OpenRA.Mods.Common.Traits.Render.WithSpriteBodyInfo info, System.Func`1[TResult] baseFacing) [0x000d6] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Mods.Common/Traits/Render/WithSpriteBody.cs:78 
  at OpenRA.Mods.Common.Traits.Render.WithSpriteBody..ctor (OpenRA.ActorInitializer init, OpenRA.Mods.Common.Traits.Render.WithSpriteBodyInfo info) [0x00000] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Mods.Common/Traits/Render/WithSpriteBody.cs:59 
  at OpenRA.Mods.RA2.Traits.Render.WithMirageSpriteBody..ctor (OpenRA.ActorInitializer init, OpenRA.Mods.RA2.Traits.Render.WithMirageSpriteBodyInfo info) [0x00000] in /home/matthias/Entwicklung/ra2/OpenRA.Mods.RA2/Traits/Render/WithMirageSpriteBody.cs:33 
  at OpenRA.Mods.RA2.Traits.Render.WithMirageSpriteBodyInfo.Create (OpenRA.ActorInitializer init) [0x00000] in /home/matthias/Entwicklung/ra2/OpenRA.Mods.RA2/Traits/Render/WithMirageSpriteBody.cs:20 
  at OpenRA.Actor..ctor (OpenRA.World world, System.String name, OpenRA.Primitives.TypeDictionary initDict) [0x000a2] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/Actor.cs:106 
  at OpenRA.World.CreateActor (System.Boolean addToWorld, System.String name, OpenRA.Primitives.TypeDictionary initDict) [0x00000] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/World.cs:331 
  at OpenRA.World.CreateActor (System.String name, OpenRA.Primitives.TypeDictionary initDict) [0x00000] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/World.cs:326 
  at OpenRA.Mods.Common.Traits.Production+<>c__DisplayClass6_0.<DoProduction>b__0 (OpenRA.World w) [0x0000d] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Mods.Common/Traits/Production.cs:82 
  at OpenRA.World.Tick () [0x0019e] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/World.cs:462 
  at OpenRA.Game.InnerLogicTick (OpenRA.Network.OrderManager orderManager) [0x001bc] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/Game.cs:621 
  at OpenRA.Game.LogicTick () [0x0003e] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/Game.cs:648 
  at OpenRA.Game.Loop () [0x000e3] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/Game.cs:795 
  at OpenRA.Game.Run () [0x0003c] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/Game.cs:835 
  at OpenRA.Game.InitializeAndRun (System.String[] args) [0x00010] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/Game.cs:263 
  at OpenRA.Program.Main (System.String[] args) [0x00044] in /home/matthias/Entwicklung/ra2/engine/OpenRA.Game/Support/Program.cs:37
```

Mailaender commented:
I changed it to only choose the tree once and not per cloak. No `MirageTarget` and no `DefaultTargetTypes` set will result in no mirage at all, which I think is a valid option for say mappers.

sorcerer86pt commented:
No trees, no mirage...... Now joking, why he couldn't mirage as a other
enemy tank, like a spy do... Or a rock....


Em 09/03/2020 à(s) 4:19 da tarde, Matthias Mailänder <
notifications@github.com> escreveu:

I changed it to only choose the tree once and not per cloak. No MirageTarget
and no DefaultTargetTypes set will result in no mirage at all, which I
think is a valid option for say mappers.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/OpenRA/ra2/pull/644?email_source=notifications&email_token=AC2N4ZZT3WARO64LHORO32TRGUJIJA5CNFSM4KZ5FQY2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEOH5MSA#issuecomment-596629064>,
or unsubscribe
<https://github.com/notifications/unsubscribe-auth/AC2N4Z626PKXP3DRSCZ7PFTRGUJIJANCNFSM4KZ5FQYQ>
.

Mailaender commented:
Mappers could add more rules to `map.yaml` and indeed have mirage tanks disguise as whatever they like, but I won't go that far here.

sorcerer86pt commented:
How the status for this one?

Mailaender commented:
It is finished.

abcdefg30 commented on OpenRA.Mods.RA2/Traits/Mirage.cs:
```diff
@@ -0,0 +1,250 @@
+			if (IsMirage)
+			{
+				wasMirage = true;
+				ActorType = targetTypes.RandomOrDefault(self.World.SharedRandom);
```
I don't think we need to do that here anymore after we did in the ctor?

Mailaender replied:
For some reason we still have to do this otherwise the sprite vanishes.

Mailaender replied:
Actually it is redundant here. The one in the constructor is critical.

abcdefg30 commented on mods/ra2/rules/allied-vehicles.yaml:
```diff
@@ -390,6 +390,19 @@ mgtk:
+		RequiresCondition: tree
+	Mirage:
+		RequiresCondition: !chronodisable
+		InitialDelay: 25
```
Why aren't those the default values?

Mailaender replied:
Good question.

abcdefg30 commented on OpenRA.Mods.RA2/Traits/Mirage.cs:
```diff
@@ -0,0 +1,250 @@
+		public readonly int InitialDelay = 10;
+
+		[Desc("Measured in game ticks.")]
+		public readonly int MirageDelay = 30;
```
Can we explain what this is in the `Desc`? I suppose this is some kind of `RevealTime`.

Mailaender replied:
I renamed it.

sorcerer86pt commented:
looks good to me. Check the video:
https://drive.google.com/file/d/1xVq4ayfoz5FICDGitpcUl2ollsc32SnC/view?usp=sharing

abcdefg30 commented on OpenRA.Mods.RA2/Traits/Render/WithMirageSpriteBody.cs:
```diff
@@ -0,0 +1,70 @@
+				{
+					var renderSprites = disguiseActor.TraitInfoOrDefault<RenderSpritesInfo>();
+					if (renderSprites != null)
+						disguiseImage = renderSprites.GetImage(disguiseActor, self.World.Map.Rules.Sequences, disguisePlayer.InternalName);
```
Actually, does this need the same fix as https://github.com/OpenRA/OpenRA/pull/17811?

Mailaender replied:
It doesn't matter here, but yes well spotted.

abcdefg30 commented on OpenRA.Mods.RA2/Traits/Render/WithMirageSpriteBody.cs:
```diff
@@ -0,0 +1,70 @@
+				if (withSpriteBody != null)
+				{
+					DefaultAnimation.PlayRepeating(NormalizeSequence(self, withSpriteBody.Sequence));
+					DefaultAnimation.ChangeImage(disguiseImage, withSpriteBody.Sequence);
```
This entire block (l59-64) should move into the `if` block above, since I suppose this will crash if `disguiseImage` is not initialized.

Mailaender replied:
Added more safeguards although you would need a serious misconfiguration to see this crash.
